### PR TITLE
go/vt/vtgate: add json and use_fallback opts to numeric_static_map vdx

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

`numeric_static_map` supports a JSON file. In some environments it may be more convenient to set this in the VSchema directly, rather than make the JSON file available everywhere it needs to be loaded.

In addition, it would be useful to have the ability to fallback to a hash lookup when there are keys missing from the json mapping.

## Related Issue(s)

 * Fixes https://github.com/vitessio/vitess/issues/12411
 * Addresses https://github.com/vitessio/vitess/issues/12412

One thing @aquarapid and I are not happy about is having to embed a JSON document inside a proto string type. If possible, we would like to amend the proto definition for vindexes to support nested, typed options. Have created https://github.com/vitessio/vitess/issues/12413 to have that discussion, and can amend this PR depending on the outcome of that.

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Documentation: https://github.com/vitessio/website/pull/1415

### CODE REVIEW

File: go/vt/vtgate/vindexes/numeric_static_map.go [hunk 1]
```diff
@@ -52,19 +53,53 @@ func init() {
...
+
+	var hashVdx Hashing
+
+	if s, ok := params["fallback_type"]; ok {
+		if err != nil {
+			return nil, err
+		}
+		// We just use xxhash blindly;  we could make this configurable.
+		vindex, err := CreateVindex(s, name+"_hash", map[string]string{})
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (maxenglander) [reply go/vt/vtgate/vindexes/numeric_static_map.go]: Done!

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/vindexes/numeric_static_map.go [hunk 2]
```diff
@@ -52,19 +53,52 @@ func init() {
...
+		lt, err = parseNumericLookupTable([]byte(jsonStr))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var hashVdx Hashing
+
+	if s, ok := params["fallback_type"]; ok {
+		if err != nil {
```
Comment (systay) [inline go/vt/vtgate/vindexes/numeric_static_map.go:87]: don't we want to check the `err` variable outside of this `if` statement?

File: go/vt/vtgate/vindexes/numeric_static_map.go [hunk 3]
```diff
@@ -121,22 +155,32 @@ func (vind *NumericStaticMap) Hash(id sqltypes.Value) ([]byte, error) {
 		return nil, err
 	}
 	lookupNum, ok := vind.lookup[num]
-	if ok {
+	if !ok {
+		// Not in lookup, us fallback hash
```
Comment (systay) [inline go/vt/vtgate/vindexes/numeric_static_map.go:159]: nitpick: `us` -> `use`

**Review by systay — APPROVED** [review]
nitpicks aside, LGTM!

File: go/vt/vtgate/vindexes/numeric_static_map.go (hunk 2 above)
Comment (maxenglander) [reply go/vt/vtgate/vindexes/numeric_static_map.go:87]: This check looks completely wrong, not sure what I was thinking here

**Review by harshit-gangal — APPROVED** [review]

**Comment by harshit-gangal** [comment]
You would need to update vindex page `https://vitess.io/docs/16.0/reference/features/vindexes/`

**Comment by harshit-gangal** [comment]
> 

It will be in 17.0 only.

**Pull request merged**